### ISSUE TITLE
Version checking before quitting app?

### ISSUE THREAD

**Issue by Theile** [scriptingosx/Installomator#11]
Not really an issue, but a feature request.

Wish that the user would not be asked to quit the given app, if the app does not need an update. Maybe it does not need to download the app either?

Could be done somehow by adding a line, that would figure out latest version of the download, if it can be scripted before download. Lates version of signal is this:
`curl -fs https://updates.signal.org/desktop/latest-mac.yml | grep version | awk '{print $2}'`

How would the admin know if the apps needs to be updated? I dream I could be lazy and run the command every morning or once per week or something, and only bug the user if the app actually needs an update.

(I realise that through Jamf Pro I can limit the target machines, but then Jamf Pro needs to know the latest version.)

**Comment by scriptingosx** [comment]
With some apps it is fairly easy to determine the latest version. With others: not so easy or even impossible. it would sure be nice, where but I am not sure I'd want to implement a check that works for only _some_ of the apps.

**Comment by scriptingosx** [comment]
I am also not convinced that _all_ clients pinging multiple random web servers _every day_ is the best approach

**Comment by Theile** [comment]
I get it, but how do you handle this in Jamf. You would have to investigate it somehow yourself, I guess… Or do you just make this available in Self Service?

When I create pkgs using AutoPkgr, it will know if it's the same version as previously downloaded, so somehow they can do it, if that is any inspiration.

**Comment by Theile** [comment]
I'm brainstorming. If it's as easy as Signal, that could be implemented and matched. If it's more difficult, the "new" version could be downloaded, and before copying a match would be made to see if the new version is actually any newer than the current, and only that would request the app to be quit.

**Comment by Theile** [comment]
My suggestion with the example of Signal.app:
Add variable to the caseStatement called appNewVersion, and the caseStatement for Signal will be as follows:
`    signal)`
`        name="Signal"`
`        type="dmg"`
`        downloadURL=https://updates.signal.org/desktop/$(curl -fs https://updates.signal.org/desktop/latest-mac.yml | grep url | grep dmg | awk '{print $3}')`
`        appName="Signal.app"`
`        appNewVersion=$(curl -fs https://updates.signal.org/desktop/latest-mac.yml | grep version | awk '{print $2}')`
`        expectedTeamID="U68MSDN6DR"`
`        ;;`

Then a fix to the mdfind command in the `getAppVersion` method/function:
`    applist=$(mdfind "kMDItemFSName='$appName' kMDItemKind='Application'" -0 )`
The double "&&" should be eliminated. At least on my system, it never finds any app.

In the main code section after "check if this is an update" and before "when user is logged in, and app is running, prompt user to quit app", insert these lines:
`# Exit if new version is the same as installed version`
`#echo "appversion $appversion" #for testing`
`#echo "appNewVersion $appNewVersion" #for testing`
`if [[ -n $appNewVersion && ($appversion == $appNewVersion) ]]; then`
`	echo "There is no new version available. Exiting."`
`	exit 0`
`fi`

So if the caseStatement does not define $appNewVersion, then it does not perform a check, but if it does and it's the sam as the version the script already finds, then it can exit.

**Comment by Theile** [comment]
Improved main code section part for debug:
`# Exit if new version is the same as installed version`
`#echo "appversion $appversion" #for testing`
`#echo "appNewVersion $appNewVersion" #for testing`
`if [[ -n $appNewVersion && ($appversion == $appNewVersion) ]]; then`
`    if [[ $DEBUG -eq 0 ]]; then`
`		echo "There is no new version available. Exiting."`
`		exit 0`
`    else`
`        echo "DEBUG mode enabled, not exiting, but there is no new version of app."`
`    fi`
`fi`

Next thought:
What if the version is only available after download of software? (I have not investigated this).

**Comment by Theile** [comment]
I have tried to make a branch and create my suggestions in there. Hope that was the right way to do it…

**Comment by isaacatmann** [comment]
This is super cool @Theile Thank You!!!  We've been watching our metrics and we see about a 60-100 "New App Version Installed Matches Existing App Version" reports per 1,000 clients for major apps (Chrome/Office/slack/zoom).  This equates to 50-70 times we probably made the user quit an active application when they were working and didn't need to.  

Putting your code though some testing with apps but adding this statement already seems to save misfires when the Jamf inventory is behind what the computer actually has and thus thinks the computer should update. With applications that have a static URL that redirects to a file with the version in it it's fairly easy to implement as well.  

Slack can be implemented with the following:

 ```
   slack)
        name="Slack"
        type="dmg"
        downloadURL="https://slack.com/ssb/download-osx"
        expectedTeamID="BQR82RBBHL"
        appNewVersion=$(curl -is https://slack.com/ssb/download-osx | grep location: | grep -o "Slack.*dmg" | cut -d "-" -f 2)
        ;;
```
This statement should work for the entire Microsoft suite in uniform, still pending testing.  
```    
microsoftword)
        name="Microsoft Word"
        type="pkg"
        downloadURL="https://go.microsoft.com/fwlink/?linkid=525134"
        expectedTeamID="UBF8T346G9"
        # msupdate requires a PPPC profile pushed out from Jamf to work, https://github.com/pbowden-msft/MobileConfigs/tree/master/Jamf-MSUpdate
        updateTool="/Library/Application Support/Microsoft/MAU2.0/Microsoft AutoUpdate.app/Contents/MacOS/msupdate"
        updateToolArguments=( --install --wait 600 --apps MSWD2019 )
        updateToolRunAsCurrentUser=yes
        appNewVersion=$(curl -is "$downloadURL" | grep ocation: | grep -o "Microsoft_.*pkg" | cut -d "_" -f 3 | cut -d "." -f1-2)
        ;;
```

We can simply make it output a nice quit message and not annoy the user as you can see in this log:

```
2020-07-29 21:40:15 INFO : https://REDACTED.jamfcloud.com/ : InstallUpdate-microsoftword : 20200726 : 5565 : We are using Jamf Self Service custom icon
2020-07-29 21:40:16 INFO : https://REDACTED.jamfcloud.com/ : InstallUpdate-microsoftword : 20200726 : 5565 : No blocking processes defined, using Microsoft Word as default
2020-07-29 21:40:16 INFO : https://REDACTED.jamfcloud.com/ : InstallUpdate-microsoftword : 20200726 : 5565 : Changing directory to /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.qINKJ3FU
2020-07-29 21:40:17 INFO : https://REDACTED.jamfcloud.com/ : InstallUpdate-microsoftword : 20200726 : 5565 : Found app at /Applications/Microsoft Word.app, version 16.39
2020-07-29 21:40:17 WARN : https://REDACTED.jamfcloud.com/ : InstallUpdate-microsoftword : 20200726 : 5565 : There is no new version available, exiting.

**Comment by Oh4sh0** [comment]
I did something similar to you. I've added the following. Note, I haven't tried subsequent releases to ensure the vendor is keeping them consistent.

iTerm2
`appNewVersion=$(curl -is https://iterm2.com/downloads/stable/latest | grep location: | grep -o "iTerm2.*zip" | cut -d "-" -f 2 | cut -d '.' -f1 | sed 's/_/./g')`

GitHub Desktop
`appNewVersion=$(curl -is https://central.github.com/deployments/desktop/desktop/latest/darwin | grep Location: | cut -d "/" -f 5 | cut -d "-" -f 1)`

Docker
`appNewVersion=$(curl -is https://docs.docker.com/docker-for-mac/release-notes/ | grep ">Docker Desktop Community" | head -1 | sed -n -e 's/^.*Community //p' | cut -d '<' -f1)`

Postman
`appNewVersion=$(curl -Is https://dl.pstmn.io/download/latest/osx | grep "content-disposition:" | sed -n -e 's/^.*Postman-osx-//p' | sed 's/\.zip//' | sed $'s/[^[:print:]\t]//g' )`

SourceTree
`appNewVersion=$(curl -fs https://product-downloads.atlassian.com/software/sourcetree/Appcast/SparkleAppcastAlpha.xml | xpath '//rss/channel/item[last()]/title' 2>/dev/null | sed -n -e 's/^.*Version //p' | sed 's/\<\/title\>//' | sed $'s/[^[:print:]\t]//g')`

Firefox
`appNewVersion=$(curl -is "https://download.mozilla.org/?product=firefox-latest&os=osx&lang=en-US" | grep Location: | grep -o "Firefox%20.*dmg" | cut -d "%" -f 2 | sed 's/20//g' | sed 's/\.dmg//g' | sed $'s/[^[:print:]\t]//g')`

Visual Studio Code
`appNewVersion=$(curl -L "https://code.visualstudio.com/Updates" | grep "/darwin" | grep -oiE ".com/([^>]+)([^<]+)/darwin" | cut -d "/" -f 2 | sed $'s/[^[:print:]\t]//g')`

VLC
`appNewVersion=$(curl -fs http://update.videolan.org/vlc/sparkle/vlc-intel64.xml | xpath '//rss/channel/item[last()]/enclosure/@sparkle:version' 2>/dev/null | cut -d '"' -f 2)`

BBEdit
`appNewVersion=$(curl https://versioncheck.barebones.com/BBEdit.xml | xpath "(//key[.='SUFeedEntryShortVersionString'])[last()]/following-sibling::string[1]/text()")`

Sublime Text
`appNewVersion=$(curl -Is https://download.sublimetext.com/latest/stable/osx | grep "Location:" | sed -n -e 's/^.*Sublime Text //p' | sed 's/.dmg//g' | sed $'s/[^[:print:]\t]//g')`

**Comment by Theile** [comment]
In my branch (https://github.com/Theile/Installomator), I have added these:
iTerm 2, Docker, Postman, SourceTree, Visual Studio Code

These, already had version check:
GitHub Desktop, Firefox, VLC, BBEdit, Sublime Text, Atom

And many others have version check, too. My version 0.4.8 is about to be released.

**Comment by Oh4sh0** [comment]
**Spotify**
`appNewVersion=$(curl https://www.spotify.com/us/opensource/ | cat | grep -o "<td>.*.</td>" | head -1 | cut -d ">" -f2 | cut -d "<" -f1)`

**VMware Horizon Client**
```
     downloadURL=$(curl "https://my.vmware.com/channel/public/api/v1.0/dlg/details?locale=en_US&downloadGroup=CART21FQ2_MAC_800&productId=1027&rPId=48989" | grep -o 'Url.*..dmg"' | cut -d '"' -f3)
	 appNewVersion=$(curl "https://my.vmware.com/channel/public/api/v1.0/dlg/details?locale=en_US&downloadGroup=CART21FQ2_MAC_800&productId=1027&rPId=48989" | sed 's/.*-\(.*\)-.*/\1/')
     expectedTeamID="EG7KH642X6"
     ;;
```

**Citrix Workspace**
```
downloadURL="https:"$(curl -s -L "https://www.citrix.com/downloads/workspace-app/mac/workspace-app-for-mac-latest.html#ctx-dl-eula-external" | grep "dmg?" | sed "s/.*rel=.\(.*\)..id=.*/\1/")
    expectedTeamID="S272Y5R93J"
    appNewVersion=$(curl https://www.citrix.com/downloads/workspace-app/mac/workspace-app-for-mac-latest.html | grep "<p>Version" | cut -d " " -f1 | cut -d ";" -f2 | cut -d "." -f 1-3)
```

**Adobe Connect**
`downloadURL="http://www.adobe.com/go/ConnectSetupMac"
    expectedTeamID="JQ525L2MZD"
    appNewVersion=$(curl https://helpx.adobe.com/adobe-connect/connect-downloads-updates.html | grep "<li>For Mac" | grep -o "/.*.dmg" | head -1 | sed 's|.*/||' | sed 's/AdobeConnect_//' | sed 's/.dmg//')`

**Comment by apizz** [comment]
I like the idea but concerned with the current implementation: https://github.com/Theile/Installomator/blob/dev/Installomator.sh#L410-L417.

The only thing that's being checked is that the currently installed version doesn't match the installed version.

What happens when the developer is slow to update their sparkle feed where installomator looks for its latest version and the user installs and newer version themselves from the website?

What happens when the developer completely changes their version scheme, or makes a tweak to it?

There has to be reasonable checks in place so that we don't inadvertently _downgrade_ a user.  We know developers (especially in macOS world) don't consistently get versioning right ... then again neither does Apple!

**Comment by Theile** [comment]
I understand this. And with the extreme long list of software, it's not easy to test all of these. I mean some test for sparkle, some try to isolate it in the name, som find it on the download page. For MS I have seen versioning behind on the XML-feed on the macadmins page, but if that was a one time situation, I don't know.

On top of that, if you look at version numbers, it's not just 10.5 compared to 10.6, so what about 10.10 is that more or less than 10.5? That makes it rather hard to compare for less or greater. And Google Chrome is version 86.0.4240.198.

Then we have the is-at-least function, and I haven't tested that. Maybe that is more intelligent, and the comparison should be if it is-at-least the new version.

When that is figured out, how do we handle reinstall, as the original version of Installomator does?

**Comment by Theile** [comment]
And I can add a version like "10.2332.565 (245.5434)" or something like that.

So that was why I ended up with just a difference. 

Please note that the download link is not dependant on version. so if the download link is correct it will always get the latest version, no matter what appNewVersion will figure out. So I do not think a newer version will ever be replaced with an older version. I would rather think that sometimes the user is asked to quit the app for an update, and the downloaded version is simply just the same.

**Comment by Oh4sh0** [comment]
The reality is there will not be a perfect solution, mainly because of there being no standard on version numbering. 

In JAMF I've taken a number of softwares, perform integer conversions of their version numbers, and that has suited fine (Chrome, Firefox, Flash, Office, etc.). Even that is problematic because there is nothing stopping devs from suddenly changing the number of digits or deciding to use letters (looking at you, Spotify) in their version numbers. 

This is more of a "best effort" approach. In a large environment, having the utility reach out and re-download the latest version of the software every time it is run is not feasible when you compound that out over thousands of users and tens of titles. So you'd want to avoid that unless there is indication that there is a newer version. But as noted, the download url doesn't rely on the version check--so you're still going out and getting the latest version in the event that something is wrong with the version number feed.. unless your download url syntax gets messed up by a revision to a vendor feed/website, etc. But then you've got that problem regardless of the versioning issue.

**Comment by Theile** [comment]
We have to investigate the log output to know what is going on. That is the option for now.

**Comment by Theile** [comment]
I have tested the above, and Spotify version is not correct. The page does not currently have the version that is actually installed by the dmg. BTW, remember “curl -fs” so it will be silent and not give failed results.

And regarding version numbers. Latest version of Spotify is “1.1.48.625.g1c87c7f7” ;-)

BTW, I have created a testing script, that will go through all labels and test if downloadURL exists and appNewVersion is giving a result with numbers. It will print what it is, rather easy (but still a manual process) to go through.

Collecting stuff for 0.4.10 release, hopefully tomorrow, Sunday.
